apis/v1alpha1/tlsroute_types.go

```diff
@@ -0,0 +1,144 @@
+	Status TLSRouteStatus `json:"status,omitempty"`
+}
+
+// TLSRouteRule is the configuration for a given rule.
+type TLSRouteRule struct {
+	// Matches define conditions used for matching the rule against
+	// incoming TLS handshake.
+	// Each match is independent, i.e. this rule will be matched
+	// if **any** one of the matches is satisfied.
+	// +optional
+	Matches []TLSRouteMatch `json:"matches"`
+	// Action defines what happens to the connection.
+	Action *TLSRouteAction `json:"action"`
+}
+
+// TLSRouteAction is the action for a given rule.
+type TLSRouteAction struct {
+	// ForwardTo sends requests to the referenced object(s). The
+	// resource may be "services" (omit or use the empty string for the
+	// group), or an implementation may support other resources (for
+	// example, resource "myroutetargets" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "services".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the HTTPRoute will be true.
+	//
+	// Support: core
+	//
+	// +kubebuilder:validation:MinItems=1
+	ForwardTo []ForwardToTarget `json:"forwardTo"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "action" behavior.  The resource may be "configmaps" (use the empty
+	// string for the group) or an implementation-defined resource (for
+	// example, resource "myrouteactions" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "configmaps".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TLSRoute will be true.
+	//
+	// Support: custom
+	//
+	// +optional
+	ExtensionRef *RouteActionExtensionObjectReference `json:"extensionRef,omitempty"`
+}
+
+// TLSRouteMatch defines the predicate used to match connections to a
+// given action.
+type TLSRouteMatch struct {
+	// SNIs defines a set of SNI names that should match against the
+	// SNI attribute of TLS CLientHello message in TLS handshake.
+	//
+	// SNI can be "precise" which is a domain name without the terminating
+	// dot of a network host (e.g. "foo.example.com") or "wildcard", which is
+	// a domain name prefixed with a single wildcard label (e.g. "*.example.com").
+	// The wildcard character '*' must appear by itself as the first DNS
+	// label and matches only a single label.
+	// You cannot have a wildcard label by itself (e.g. Host == "*").
+	// Requests will be matched against the Host field in the following order:
+	//
+	// 1. If SNI is precise, the request matches this rule if
+	//    the SNI in ClientHello is equal to one of the defined SNIs.
+	// 2. If SNI is a wildcard, then the request matches this rule if
+	//    the SNI is to equal to the suffix
+	//    (removing the first label) of the wildcard rule.
+	//
+	// Support: core
+	//
```

Comment:
Nit: we should define min and max lengths here. Maybe 1 and 10 respectively?

Reply:
Can we do this in another PR all over the project as part of #290?
This requires defining a new type (aliasing string) and then defining the length restrictions on that type.

I've added the following for the array length validation:
	// +kubebuilder:validation:MinItems=1

Reply:
Oh good catch, I was actually meaning min and max lengths of the list, though both string and list would be good. Mind adding a kubebuilder:validation:MaxItems line here?

Reply:
I'm fine with waiting for the string length restriction

Reply:
> Mind adding a kubebuilder:validation:MaxItems line here?

Updated.